File: contract/assembly/models/Recipe.ts
import { AccountID, getCurrentDate, getID } from "../utils";
import Image from "./Image";
import Ingridient from "./Ingridient";

@nearBindgen
class Recipe {
  id: string;
  recipeBookID: string;
  image: Image;
  creator: AccountID;
  category: string;
  title: string;
  description: string;
  chefNote: string;
  ingredients: Array<Ingridient>;
  instructions: Array<string>;
  reviews: Array<string>;
  ratings: Array<f64>;
  averageRating: f64;
  totalTips: f64;
  createdAt: string;

  constructor(
    creator: AccountID,
    title: string,
    description: string,
    ingridients: Array<Ingridient>,
    instructions: Array<string>,
    recipeBookID: string,
    category: string,
    chefNote: string
  ) {
    this.id = getID();
    this.recipeBookID = recipeBookID;
    this.image = new Image("", "", "");
    this.creator = creator;
    this.category = category;
    this.title = title;
    this.description = description;
    (this.chefNote = chefNote), (this.ingredients = ingridients);
    this.instructions = instructions;
    this.reviews = new Array();
    this.ratings = new Array();
    this.averageRating = 0;
    this.totalTips = 0;
    this.createdAt = getCurrentDate();
  }

  // set recipe title
  setTitle(title: string): void {
    this.title = title;
  }
  
  // set recipe description
  setDescription(description: string): void {
    this.description = description;
  }

  // sets the recipe image banner.
  setImage(image: Image | null = null): void {
    if(image) {
      this.image = new Image(image.name, image.cid, image.url);
    } else {
      this.image = new Image("", "", "");
    }
  }

  // sets the total of tips given to the recipe creator
  setTotalTips(totalTips: number): void {
    this.totalTips = totalTips;
  }

  // sets category of recipe
  setCategory(category: string): void {
    this.category = category;
  }

  // sets chefNote for recipe
  setChefNote(chefNote: string): void {
    this.chefNote = chefNote;
  }

  // Adds ingridient to ingridients.
  addIngridient(
    label: string,
    amount: i32,
    unit: string,
    details: string
  ): void {
    this.ingredients.push(new Ingridient(label, amount, unit, details));
  }

  // Removes ingridient from ingridients.
  removeIngridient(ingredientID: string): void {
    let index = -1;

    for (let i = 0; i < this.ingredients.length; i++) {
      if (this.ingredients[i].id == ingredientID) index = i;
    }

    assert(index, "Ingredient not found.");

    this.ingredients.splice(index, 1);
  }

  // Adds step to instructions.
  addStep(step: string): void {
    this.instructions.push(step);
  }
  // Removes step from instructions.
  removeStep(step: string): void {
    const index = this.instructions.indexOf(step);
    this.instructions.splice(index, 1);
  }

  // adds new review
  addReview(reviewID: string): void {
    this.reviews.push(reviewID);
  }

  // add new Rating.
  addRating(rating: f64): void {
    this.ratings.push(rating);
  }

  // updates average raiting.
  updateAverageRating(): void {
    let timesRated = this.ratings.length;
    let ratingTotal: f64 = 0;

    for (let i = 0; i < this.ratings.length; i++) {
      ratingTotal = ratingTotal + this.ratings[i];
    }

    // Divide if timesRated is not zero.
    if (timesRated === 0 || isNaN(timesRated)) {
      this.averageRating = 0;
    }
    // else return 0 as this.averageRating.
    else {
      this.averageRating = ratingTotal / timesRated;
    }
  }

  deleteReview(id: string): void {
    this.reviews.splice(this.reviews.indexOf(id), 1);
  }

  deleteRating(rating: f64): void {
    this.ratings.splice(this.ratings.indexOf(rating), 1);
  }
}

export default Recipe;

File: contracts/rust-contract/src/structs/Recipe.rs
use super::image::Image;
use super::ingredient::Ingredient;
use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
use near_sdk::serde::{Deserialize, Serialize};
use near_sdk::AccountId;

#[derive(Serialize, Deserialize, BorshDeserialize, BorshSerialize)]
#[serde(crate = "near_sdk::serde")]
pub struct Recipe {
    pub id: i128,
    pub recipe_book_id: i128,
    pub image: Image,
    pub creator: AccountId,
    pub category: String,
    pub title: String,
    pub description: String,
    pub chef_note: String,
    pub ingredients: Vec<Ingredient>,
    pub instructions: Vec<String>,
    pub reviews: Vec<String>,
    pub ratings: Vec<f64>,
    pub average_rating: f64,
    pub total_tips: f64,
    pub created_at: String,
}

impl Recipe {
    pub fn update_average_rating(&mut self) {
        let times_rated = *&self.ratings.len() as f64;
        let mut rating_total: f64 = 0.0;

        for rating in &*&self.ratings {
            rating_total = rating_total + rating;
        }

        // Divide if times_rated is not zero.
        if times_rated == 0.0 {
            self.average_rating = 0.0;
        }
        // else return 0 as this.averageRating.
        else {
            self.average_rating = rating_total / times_rated;
        }
    }

    pub fn delete_rating(&mut self, rating: f64) {
        self.ratings = self
            .ratings
            .iter()
            .filter(|x| x == &&rating)
            .map(|x| x)
            .cloned()
            .collect();
    }
}
